### DISCUSSION TITLE
設定資料表欄位 UNIQUE 屬性時遇到 MySQL ERROR 1067: Invalid default value for ‘created_at’ 的錯誤

### DISCUSSION THREAD

**Discussion by outshaker** [Lidemy/forum#146] [後端相關問題]
因為一開始建立留言板 users 資料表的時候沒有設定到 username 的 UNIQUE 屬性，然後使用的資料庫管理工具是 Adminer 4.8.1，不確定要在哪邊設定，所以查了一下用 SQL 指令語法。

```sql
ALTER TABLE  users ADD UNIQUE (username) --設定 username 在表格的唯一屬性
```

然後執行上面這行指令的時候跳出 MySQL ERROR 1067: Invalid default value for ‘created_at’ 的錯誤訊息

根據關鍵字查詢找到一篇文章：[mysql - ERROR 1067 (42000): Invalid default value for 'created_at' - Stack Overflow](https://stackoverflow.com/questions/36882149/error-1067-42000-invalid-default-value-for-created-at)

以下是根據這篇文章提供的資訊解決問題

**Comment by outshaker** [comment]
1. 文章說這錯誤是因為兩個資料庫的設定值造成的 [(參考說明)](https://www.azureunali.com/dbmysql-sql-mode%E7%9A%84%E8%A8%AD%E5%AE%9A/ "[DB][MySQL] SQL-Mode的設定 – 蒼月之嵐")
  - NO_ZERO_IN_DATE 嚴格模式時，不允許日期和月份有零
  - NO_ZERO_DATE 嚴格模式時，不允許增加零日期，會跳出錯誤而不是警告

2. 確認自己資料庫狀態是否符合？
  有的資料庫管理工具可能找不到這功能，打指令查可能比較快
  ```sql
  select @@sql_mode;
  ```
  結果得到 **NO_ZERO_IN_DATE,NO_ZERO_DATE,NO_ENGINE_SUBSTITUTION** 表示確實是這個原因導致的錯誤

3. 文章的提到的解決方式：
  - 把 @@sql_mode 裡面的設定移除：
    簡單粗暴，但參考文章內的網友建議線上環境不要這樣子做
  - 修正造成錯誤的資料欄位設定：
    預設值不要用空日期，下面說明修正問題的 SQL 指令

4. 選擇修正日期欄位設定的作法 [(官方文件詳細說明)](https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html "MySQL :: MySQL 8.0 Reference Manual :: 11.2.5 Automatic Initialization and Updating for TIMESTAMP and DATETIME")

```sql
ALTER TABLE `users` MODIFY created_at DATETIME DEFAULT CURRENT_TIMESTAMP NOT NULL;
```

## 後記

### 在 Adminer 下設定 created_at 欄位自動帶入當前時間
導致問題的原因是在 Adminer 4.8.1 設定日期預設值的方式跟 phpMyadmin 有點不太一樣，以下參照附圖說明問題成因和正確做法。

錯誤示範
![建立資料表 (錯誤)](https://user-images.githubusercontent.com/536861/122493211-dfa7ad80-d019-11eb-9f10-17921e88b496.jpg)
![欄位設定 (錯誤)](https://user-images.githubusercontent.com/536861/122493216-e0d8da80-d019-11eb-8b18-c68137a2dd2a.jpg)
上面是在嚴格模式下會觸發錯誤

正確的設定方式
![資料欄位設定 (正確)](https://user-images.githubusercontent.com/536861/122493217-e0d8da80-d019-11eb-87d4-06285db39d57.jpg)
![欄位設定 (正確)](https://user-images.githubusercontent.com/536861/122493219-e1717100-d019-11eb-9c90-3a5919cf3f40.jpg)

### 在 Adminer 下設定資料欄位的 UNIQUE 屬性
另外在 Adminer 設定資料欄位的唯一屬性，要透過**修改索引**的方式，Adminer 把 PRIMARY, UNIQUE, INDEX 這一類的欄位屬性設定都歸類在索引裡面。

![設定 UNIQUE-1](https://user-images.githubusercontent.com/536861/122493220-e20a0780-d019-11eb-8626-8d9b2d97f579.jpg)
![設定 UNIQUE-2](https://user-images.githubusercontent.com/536861/122493221-e20a0780-d019-11eb-8d2a-86c91910529e.jpg)

然後如果是用其他套 GUI 找不到設定 UNIQUE 地方的話，最後大絕招就是下 SQL 指令 `ALTER TABLE  users ADD UNIQUE (username)` 。

# 附錄：用純 cli 指令檢查、修正問題

```sql
-- 登入資料庫伺服器
C:\WINDOWS\system32>mysql -u root -p
Enter password: ***
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 530
Server version: 10.4.19-MariaDB mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

-- 選擇要使用的資料庫
MariaDB [six]> use demo
Database changed

-- 查詢資料表欄位設定
MariaDB [demo]> show create table users;
+-------+----------------------------------------------+
| Table | Create Table|                                |
+-------+----------------------------------------------+
| users | CREATE TABLE `users` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `nickname` varchar(64) NOT NULL,
  `username` varchar(64) NOT NULL,
  `pass` varchar(128) NOT NULL,
  `created_at` datetime NOT NULL DEFAULT '0000-00-00 00:00:00' ON UPDATE current_timestamp(),
  -- 確認是有問題的設定
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=18 DEFAULT CHARSET=utf8 |
+-------+----------------------------------------------+
1 row in set (0.001 sec)

-- 輸入修正的指令
MariaDB [demo]> ALTER TABLE `users` MODIFY created_at DATETIME DEFAULT CURRENT_TIMESTAMP NOT NULL;
Query OK, 0 rows affected (0.036 sec)
Records: 0  Duplicates: 0  Warnings: 0

-- 確認設定有成功修改
MariaDB [demo]> show create table users;
+-------+----------------------------------------------+
| Table | Create Table                                 |
+-------+----------------------------------------------+
| users | CREATE TABLE `users` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `nickname` varchar(64) NOT NULL,
  `username` varchar(64) NOT NULL,
  `pass` varchar(128) NOT NULL,
  `created_at` datetime NOT NULL DEFAULT current_timestamp(),
  -- 這邊預設值就會自動帶入現在的時間
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=18 DEFAULT CHARSET=utf8 |
+-------+----------------------------------------------+
1 row in set (0.000 sec)

-- 輸入一開始原本要打的指令
MariaDB [demo]> ALTER TABLE  users ADD UNIQUE (username);
Query OK, 0 rows affected, 1 warning (0.425 sec)
Records: 0  Duplicates: 0  Warnings: 1
```

**Reply by ddylanlinn** [reply]
感謝分享
另外想問一下用 CLI 的優點是什麼呀

**Reply by outshaker** [reply]
課程後期在雲端部署通常都是用 cli 下指令
而且 gui 通常只會提供常用功能，像 `select @@sql_mode;` 想要查詢資料庫內部設定參數的時候就沒辦法透過 gui 完成，至少我目前使用的 adminer 就找不到
